# Steam Game Popularity Modeling

In [36]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Data Cleaning

We will clean the table by selecting only the columns that we need for our model.

In [37]:
df = pd.read_csv("../data/a_steam_data_2021_2025.csv")
df = df[['release_year', 'genres', 'categories', 'price', 'developer', 'publisher', 'recommendations']]
df

release_year                                          genres  \
0              2024             Action;Adventure;Indie;RPG;Strategy   
1              2025                Casual;Indie;Simulation;Strategy   
2              2025                        Adventure;Indie;Strategy   
3              2023                          Action;Adventure;Indie   
4              2021  Action;Adventure;Casual;Indie;RPG;Early Access   
...             ...                                             ...   
65516          2024                         Casual;Indie;Simulation   
65517          2025                       Action;Indie;Early Access   
65518          2021                            Simulation;Education   
65519          2024                                    Action;Indie   
65520          2021                                    Casual;Indie   

                                              categories  price  \
0                           Single-player;Family Sharing   3.99   
1              Single-player;Save Anytime;Family Sharing   7.99   
2      Single-player;Steam Achievements;Full controll...  12.99   
3      Single-player;Steam Achievements;Full controll...  24.99   
4      Single-player;Steam Achievements;Full controll...   3.99   
...                                                  ...    ...   
65516                       Single-player;Family Sharing  14.99   
65517                       Single-player;Family Sharing   4.99   
65518  Single-player;Tracked Controller Support;VR On...  14.99   
65519  Single-player;Partial Controller Support;Famil...  15.99   
65520                       Single-player;Family Sharing   0.99   

                   developer              publisher  recommendations  
0      Bouncy Rocket Studios  Bouncy Rocket Studios                0  
1           Caramelo Studios       Caramelo Studios                0  
2                 Libragames             Libragames                0  
3                     YS Net                   ININ                0  
4           Wagner Rodrigues       Wagner Rodrigues                0  
...                      ...                    ...              ...  
65516           Inonia Games           Inonia Games                0  
65517       Hard Shark Games       Hard Shark Games                0  
65518        Xennial Digital        Xennial Digital                0  
65519            PoisonGames            PoisonGames                0  
65520           Rylan Yancey           Rylan Yancey                0  

[65521 rows x 7 columns]

Although certain predictor fields display some missing data, we will simply treat them accordingly in our model rather than dropping them. Multi-label fields such as `genres` and `categories` allow empty values, and missing `developer` and `publisher` will be treated as a separate category. Since no missing data is seen for recommendations, we will not be dropping any rows entirely.

## Target Construction

First, we create a new scalar containing all the values for non-zero recommendations count, and find the 0.5 and 0.8 quantile to group our observations into their popularity tier.

In [61]:
non_zero = df.loc[df["recommendations"] > 0, "recommendations"]
q50 = float(non_zero.quantile(0.50))
q80 = float(non_zero.quantile(0.80))


We then define our new 'popularity' field, with numerical values 0, 1, 2 and 3 corresponding to the tier categories.

In [71]:
conditions = [
    (df['recommendations'] > 0) & (df['recommendations'] < q50),
    (df['recommendations'] >= q50) & (df['recommendations'] < q80),
    df['recommendations'] >= q80
]

choices = [ 1, 2, 3]

df['popularity'] = np.select(conditions, choices, default = 0)
df

release_year                                          genres  \
0              2024             Action;Adventure;Indie;RPG;Strategy   
1              2025                Casual;Indie;Simulation;Strategy   
2              2025                        Adventure;Indie;Strategy   
3              2023                          Action;Adventure;Indie   
4              2021  Action;Adventure;Casual;Indie;RPG;Early Access   
...             ...                                             ...   
65516          2024                         Casual;Indie;Simulation   
65517          2025                       Action;Indie;Early Access   
65518          2021                            Simulation;Education   
65519          2024                                    Action;Indie   
65520          2021                                    Casual;Indie   

                                              categories  price  \
0                           Single-player;Family Sharing   3.99   
1              Single-player;Save Anytime;Family Sharing   7.99   
2      Single-player;Steam Achievements;Full controll...  12.99   
3      Single-player;Steam Achievements;Full controll...  24.99   
4      Single-player;Steam Achievements;Full controll...   3.99   
...                                                  ...    ...   
65516                       Single-player;Family Sharing  14.99   
65517                       Single-player;Family Sharing   4.99   
65518  Single-player;Tracked Controller Support;VR On...  14.99   
65519  Single-player;Partial Controller Support;Famil...  15.99   
65520                       Single-player;Family Sharing   0.99   

                   developer              publisher  recommendations  \
0      Bouncy Rocket Studios  Bouncy Rocket Studios                0   
1           Caramelo Studios       Caramelo Studios                0   
2                 Libragames             Libragames                0   
3                     YS Net                   ININ                0   
4           Wagner Rodrigues       Wagner Rodrigues                0   
...                      ...                    ...              ...   
65516           Inonia Games           Inonia Games                0   
65517       Hard Shark Games       Hard Shark Games                0   
65518        Xennial Digital        Xennial Digital                0   
65519            PoisonGames            PoisonGames                0   
65520           Rylan Yancey           Rylan Yancey                0   

       popularity  
0               0  
1               0  
2               0  
3               0  
4               0  
...           ...  
65516           0  
65517           0  
65518           0  
65519           0  
65520           0  

[65521 rows x 8 columns]

## Data Splitting
The data will be split chronologically, where older games are used for our training set and newer ones for our testing set. This is because a time-based split is more practical for generalizing data over time. We can inspect the proportion for each release year to decide the cutoff.

In [72]:
df['release_year'].value_counts(normalize=True)

release_year
2025    0.306863
2024    0.242029
2023    0.176386
2022    0.146777
2021    0.127944
Name: proportion, dtype: float64

Since games released in 2025 takes up roughly 30.7% of all observations, we will be using all games during 2021-2024 for training and those in 2025 for testing for a 70-30 split.

In [68]:
training = df[df['release_year'] != 2025]
testing = df[df['release_year'] == 2025]
training

release_year                                          genres  \
0              2024             Action;Adventure;Indie;RPG;Strategy   
3              2023                          Action;Adventure;Indie   
4              2021  Action;Adventure;Casual;Indie;RPG;Early Access   
5              2023                          Action;Adventure;Indie   
11             2024                      Action;Adventure;Indie;RPG   
...             ...                                             ...   
65515          2023                                          Casual   
65516          2024                         Casual;Indie;Simulation   
65518          2021                            Simulation;Education   
65519          2024                                    Action;Indie   
65520          2021                                    Casual;Indie   

                                              categories  price  \
0                           Single-player;Family Sharing   3.99   
3      Single-player;Steam Achievements;Full controll...  24.99   
4      Single-player;Steam Achievements;Full controll...   3.99   
5      Single-player;Full controller support;Family S...   1.99   
11     Single-player;Steam Achievements;Full controll...  14.99   
...                                                  ...    ...   
65515                       Single-player;Family Sharing   6.99   
65516                       Single-player;Family Sharing  14.99   
65518  Single-player;Tracked Controller Support;VR On...  14.99   
65519  Single-player;Partial Controller Support;Famil...  15.99   
65520                       Single-player;Family Sharing   0.99   

                   developer                   publisher  recommendations  \
0      Bouncy Rocket Studios       Bouncy Rocket Studios                0   
3                     YS Net                        ININ                0   
4           Wagner Rodrigues            Wagner Rodrigues                0   
5                 Opia Games  Opia Games;Plug In Digital                0   
11               Sam Enright                United Label              240   
...                      ...                         ...              ...   
65515      Seven Sails Games           Seven Sails Games                0   
65516           Inonia Games                Inonia Games                0   
65518        Xennial Digital             Xennial Digital                0   
65519            PoisonGames                 PoisonGames                0   
65520           Rylan Yancey                Rylan Yancey                0   

       popularity  
0               0  
3               0  
4               0  
5               0  
11              1  
...           ...  
65515           0  
65516           0  
65518           0  
65519           0  
65520           0  

[45415 rows x 8 columns]

In [69]:
testing

release_year                                             genres  \
1              2025                   Casual;Indie;Simulation;Strategy   
2              2025                           Adventure;Indie;Strategy   
6              2025                          Action;Indie;Free To Play   
7              2025                                       Action;Indie   
8              2025                             Adventure;RPG;Strategy   
...             ...                                                ...   
65492          2025  Adventure;Casual;Massively Multiplayer;RPG;Sim...   
65494          2025                                           Strategy   
65500          2025           Action;Adventure;Casual;Indie;Simulation   
65511          2025                             Adventure;Casual;Indie   
65517          2025                          Action;Indie;Early Access   

                                              categories  price  \
1              Single-player;Save Anytime;Family Sharing   7.99   
2      Single-player;Steam Achievements;Full controll...  12.99   
6      Single-player;Full controller support;Custom V...   0.00   
7      Single-player;Steam Achievements;Full controll...   2.99   
8                           Single-player;Family Sharing   3.99   
...                                                  ...    ...   
65492  Multi-player;MMO;PvP;Online PvP;Co-op;Online C...   0.00   
65494  Single-player;Steam Achievements;In-App Purcha...   4.99   
65500  Single-player;Steam Achievements;Full controll...   4.99   
65511                       Single-player;Family Sharing   1.99   
65517                       Single-player;Family Sharing   4.99   

                developer           publisher  recommendations  popularity  
1        Caramelo Studios    Caramelo Studios                0           0  
2              Libragames          Libragames                0           0  
6         Taylor Conolley     Taylor Conolley                0           0  
7              halvardo13          halvardo13                0           0  
8               Oba Games           Oba Games                0           0  
...                   ...                 ...              ...         ...  
65492        Vitali Kirpu        Vitali Kirpu                0           0  
65494           SHN Games           SHN Games                0           0  
65500  Jon Warner Studios  Jon Warner Studios                0           0  
65511               Total               Total                0           0  
65517    Hard Shark Games    Hard Shark Games                0           0  

[20106 rows x 8 columns]

## Feature Engineering and Encoding

In [75]:
training['popularity'].value_counts()

popularity
0    39477
1     2928
2     1772
3     1238
Name: count, dtype: int64

The popularity tiers are clearly imbalanced, with Tier 0 dominating the dataset. This is expected, since many Steam games receive little or no user engagement. Among games that do receive recommendations, the remaining tiers are reasonably balanced due to the quantile-based split.

Rather than increasing the representation of higher-popularity tiers, this project's approach will be to reduce the number of Tier 0 observations in the testing set. Oversampling higher tiers could introduce noise, particularly given the high-cardinality nature of features such as developer and publisher, which may lead to unstable patterns. Undersampling Tier 0 helps mitigate imbalance while preserving the relative structure among engaged games.

We will be keeping Tier 0 the same size as all other tiers combined.

In [85]:
tier_0 = training[training['popularity'] == 0]
tier_rest = training[training['popularity'] != 0]

n_keep = len(tier_rest)

tier_0_sampled = tier_0.sample(
    n=n_keep,
    random_state=17
)

# Combining the undersampled tier 0 dataframe and the original dataframe for all other tiers
undersampled = pd.concat([tier_0_sampled, tier_rest])

# Shuffle all observations
undersampled = undersampled.sample(frac=1, random_state=17).reset_index(drop=True)
undersampled

release_year                                genres  \
0              2024                          Action;Indie   
1              2024   Adventure;Casual;Indie;Free To Play   
2              2023                          Casual;Indie   
3              2023                Adventure;Casual;Indie   
4              2023      Casual;Indie;Simulation;Strategy   
...             ...                                   ...   
11871          2024  Action;Adventure;Casual;RPG;Strategy   
11872          2021                          Casual;Indie   
11873          2023                                 Indie   
11874          2021                   Action;Casual;Indie   
11875          2023        Action;Casual;Indie;Simulation   

                                              categories  price  \
0      Single-player;Partial Controller Support;Famil...   9.99   
1                                          Single-player   0.00   
2                           Single-player;Family Sharing   0.99   
3      Single-player;Steam Achievements;Steam Trading...  19.99   
4      Single-player;Steam Achievements;Steam Trading...   0.99   
...                                                  ...    ...   
11871  Single-player;Steam Achievements;Full controll...   2.99   
11872  Single-player;Steam Achievements;Steam Trading...   9.99   
11873                       Single-player;Family Sharing   1.75   
11874  Single-player;Multi-player;PvP;Shared/Split Sc...   4.99   
11875  Single-player;Steam Achievements;Full controll...  12.99   

                    developer               publisher  recommendations  \
0         Little Crows Studio     Little Crows Studio                0   
1                    Ebananas               ebanapury                0   
2                Sarni Studio            Sarni Studio                0   
3                    Bad Seed   Maximum Entertainment              356   
4               Watashi Games           Watashi Games             1715   
...                       ...                     ...              ...   
11871        Soda Game Studio        Soda Game Studio              372   
11872  Business Goose Studios  Business Goose Studios             2502   
11873              Lewd Leech              Lewd Leech                0   
11874   One-Hand-Free Studios   One-Hand-Free Studios                0   
11875            Neko Machine            Neko Machine              102   

       popularity  
0               0  
1               0  
2               0  
3               2  
4               3  
...           ...  
11871           2  
11872           3  
11873           0  
11874           0  
11875           1  

[11876 rows x 8 columns]


In this context, predicting a game as highly popular when it ultimately performs poorly is considered more costly than the opposite error. A false positive may lead to overestimating success and misallocating effort, whereas a false negative still allows room for iteration and improvement after release. For this reason, the modeling approach is intentionally conservative when identifying high-popularity games, and false positives are treated as more costly than false negatives.